cmd/bad-key-revoker/main.go

```diff
@@ -461,29 +472,74 @@ func main() {
 	cmd.FailOnError(err, fmt.Sprintf("failed to parse email template %q: %s", config.BadKeyRevoker.Mailer.EmailTemplate, err))
 
 	bkr := &badKeyRevoker{
-		dbMap:           dbMap,
-		maxRevocations:  config.BadKeyRevoker.MaximumRevocations,
-		serialBatchSize: config.BadKeyRevoker.FindCertificatesBatchSize,
-		raClient:        rac,
-		mailer:          mailClient,
-		emailSubject:    config.BadKeyRevoker.Mailer.EmailSubject,
-		emailTemplate:   emailTemplate,
-		logger:          logger,
-		clk:             clk,
+		dbMap:               dbMap,
+		maxRevocations:      config.BadKeyRevoker.MaximumRevocations,
+		serialBatchSize:     config.BadKeyRevoker.FindCertificatesBatchSize,
+		raClient:            rac,
+		mailer:              mailClient,
+		emailSubject:        config.BadKeyRevoker.Mailer.EmailSubject,
+		emailTemplate:       emailTemplate,
+		logger:              logger,
+		clk:                 clk,
+		backoffDurationMax:  config.BadKeyRevoker.BackoffDurationMax.Duration,
+		backoffDurationBase: config.BadKeyRevoker.Interval.Duration,
+		backoffFactor:       1.3,
+	}
+
+	// If `BackoffDurationMax` was not set via the config, set it to 60
+	// seconds. This will avoid a tight loop on error but not be an
+	// excessive delay if the config value was not deliberately set.
+	if config.BadKeyRevoker.BackoffDurationMax.Duration == 0 {
+		bkr.backoffDurationMax = time.Second * 60
 	}
+
+	// If `Interval` is not set in the config then set
+	// `bkr.backoffDurationBase` to a default 1 second.
+	if config.BadKeyRevoker.Interval.Duration == 0 {
+		bkr.backoffDurationBase = time.Second
+	}
+
+	// Run bad-key-revoker in a loop. Backoff if no work or errors.
 	for {
 		noWork, err := bkr.invoke()
 		if err != nil {
 			keysProcessed.WithLabelValues("error").Inc()
 			logger.AuditErrf("failed to process blockedKeys row: %s", err)
+			// Increase backoff duration.
+			bkr.backoffTick()
+			logger.Infof("backoff on error, trying again in %f seconds.", bkr.backoffDuration.Seconds())
+			// Sleep for backoff duration.
+			bkr.clk.Sleep(bkr.backoffDuration)
 			continue
 		}
 		if noWork {
-			logger.Info(fmt.Sprintf(
-				"No work to do. Sleeping for %s", config.BadKeyRevoker.Interval.Duration))
-			time.Sleep(config.BadKeyRevoker.Interval.Duration)
+			// Increase backoff duration.
+			bkr.backoffTick()
+			logger.Infof("no work to do, trying again in %f seconds", bkr.backoffDuration.Seconds())
+			// Sleep for `backoffDuration`
+			bkr.clk.Sleep(bkr.backoffDuration)
 		} else {
 			keysProcessed.WithLabelValues("success").Inc()
+			// Successfully processed, reset backoff.
+			bkr.backoffReset()
 		}
 	}
 }
+
+// backoffTick increments the backoffTicker and calls core.RetryBackoff
+// to calculate a new backoff duration and set `bkr.backoffDuration`.
+func (bkr *badKeyRevoker) backoffTick() {
+	bkr.backoffTicker++
+	bkr.backoffDuration = core.RetryBackoff(
+		bkr.backoffTicker,
+		bkr.backoffDurationBase,
+		bkr.backoffDurationMax,
+		bkr.backoffFactor,
+	)
+}
```

Comment:
As per my comment above, I'd change this to be something like:

```suggestion
func (bkr *badKeyRevoker) backoff() {
	brk.clk.Sleep(brk.backoffDuration)
	bkr.backoffTicker++
	bkr.backoffDuration = core.RetryBackoff(
		bkr.backoffTicker,
		bkr.backoffDurationBase,
		bkr.backoffDurationMax,
		bkr.backoffFactor,
	)
}
```

This also ensures that the first time we ever sleep, we sleep for exactly the minimum interval, instead of incrementing the ticker and calculating a new backoff duration _before_ that very first sleep.

Reply:
Yeah, I can add the sleep to the function. We still need to increment the backoffTicker first because a 0 ticker passed to core.RetryBackoff returns a 0 duration. So in this case we are indeed sleeping for the minimum interval the first time, except, of course, plus or minus the jitter in the calculation.

I think I could actually get rid of bkr.backoffDuration altogether and put the calculation inline with the sleep.

Reply:
Updated to include them all in the method and used a generic backoff message for the log. This makes the test a bit less useful because I am now only checking that the backoff is logged since it is harder to get the backoff duration in the tests. I suppose `core.RetryBackoff` already has tests so I don't really need to check the algorithm.